### PR TITLE
Follow alert

### PR DESCRIPTION
**Pull request by Specture7**
Gives the option to disable and enable a new follower alert

### CODE REVIEW

**Comment by Top-Cat** [comment]
ok to test

File: src/jvmMain/kotlin/io/beatmaps/api/users.kt [hunk 1]
```diff
@@ -919,21 +920,30 @@ fun Route.userRoute() {
...
             transaction {
-                if (req.following) {
-                    Follows.upsert(conflictIndex = Follows.link) { follow ->
-                        follow[userId] = req.userId
-                        follow[followerId] = user.userId
-                        follow[since] = NowExpression(since.columnType)
-                        follow[upload] = req.upload
-                        follow[curation] = req.curation
-                    }
+                val shouldAlert = Follows.select { Follows.userId eq req.userId }.empty()
```
Comment (Top-Cat) [inline src/jvmMain/kotlin/io/beatmaps/api/users.kt:923]: This condition checks if **anyone** follows the user in question

File: src/jvmMain/kotlin/io/beatmaps/api/users.kt [hunk 2]
```diff
@@ -919,21 +920,30 @@ fun Route.userRoute() {
...
+
+                Follows.upsert(conflictIndex = Follows.link) { follow ->
+                    follow[userId] = req.userId
+                    follow[followerId] = user.userId
+                    follow[since] = NowExpression(since.columnType)
+                    follow[upload] = req.upload
+                    follow[curation] = req.curation
+                    follow[following] = req.following
+                }
+                val followedUser = if (shouldAlert) {
```
Comment (Top-Cat) [inline src/jvmMain/kotlin/io/beatmaps/api/users.kt:933]: This nullable into null coalescing seems a bit too esoteric to me.
Just wrap the whole thing in the condition...

```kt
if (shouldAlert) {
    val followedUser = UserDao.wrapRow(User.select { User.id eq req.userId }.single())

    if (followedUser.followAlerts) {
        Alert.insert(
            "New Follower",
            "@${user.uniqueName} is now following you!",
            EAlertType.Follow,
            req.userId
        )
    }
}
```

(single should also be safe enough as the insert should fail if the user id is invalid)

File: src/jvmMain/kotlin/io/beatmaps/api/users.kt [hunk 3]
```diff
@@ -919,21 +920,30 @@ fun Route.userRoute() {
...
+                    User.select { User.id eq req.userId }.first().let { r -> UserDao.wrapRow(r) }
                 } else {
-                    Follows.deleteWhere {
-                        (Follows.userId eq req.userId) and (Follows.followerId eq user.userId)
-                    }
+                    null
+                }
+                if (followedUser?.followAlerts == true) {
+                    Alert.insert(
```
Comment (Top-Cat) [inline src/jvmMain/kotlin/io/beatmaps/api/users.kt:939]: I think you need to call `updateAlertCount(req.userId)` after Alert.insert

**Comment by Specture7** [comment]
problems should be fixed now

**Comment by Specture7** [comment]
re-added the bracket that was missing before

**Pull request merged**